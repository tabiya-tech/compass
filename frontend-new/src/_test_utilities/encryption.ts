import { webcrypto as crypto } from "crypto";
import { EncryptionConfig } from "src/sensitiveData/config/encryptionConfig";

// Constraining the key size to 2048 or 4096 bits for RSA as the backend expect the 4096 in calculating the maximum size of the encrypted key.
export async function generateRSACryptoPairKey(keySize: 2048 | 4096) {
  return await crypto.subtle.generateKey(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
      modulusLength: keySize,
      // 3 is a common exponent in practice
      publicExponent: new Uint8Array([1, 0, 1]),
      hash: EncryptionConfig.RSA.KEY_HASH_FN,
    },
    true, // Whether the key is extractable (i.e. can be used in exportKey)
    ["encrypt", "decrypt"] // the key can be used for both encryption and decryption
  );
}

function arrayBufferToBase64(buffer: ArrayBuffer): string {
  // @ts-ignore
  const binary = String.fromCharCode(...new Uint8Array(buffer));

  return Buffer.from(binary, "binary").toString("base64");
}

/**
 * Export a CryptoKey to a PEM formatted string (public key)
 *
 * @param key The CryptoKey to export as a PEM formatted string
 * @returns The PEM formatted string.
 */
export async function exportCryptoPublicKey(key: CryptoKey) {
  // Export the key to ArrayBuffer
  const exported = await crypto.subtle.exportKey(EncryptionConfig.RSA.FORMAT, key);

  const base64String = arrayBufferToBase64(exported);

  // add the PEM header and footer to make it a PEM formatted string.
  return `-----BEGIN PUBLIC KEY-----\n${base64String}\n-----END PUBLIC KEY-----`;
}

/**
 * Decrypt the encrypted data with the private key
 * using RSA Algorithm.
 *
 * @param privateKey
 * @param encryptedData
 */
export async function decryptWithRSA(privateKey: CryptoKey, encryptedData: ArrayBuffer) {
  return await crypto.subtle.decrypt(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
    },
    privateKey,
    encryptedData
  );
}

export async function decryptWithAES(encryptedData: ArrayBuffer, iv: ArrayBuffer, key: ArrayBuffer) {
  try {
    const importedKey = await crypto.subtle.importKey(
      EncryptionConfig.AES.FORMAT,
      key,
      EncryptionConfig.AES.ALGORITHM,
      true,
      ["decrypt"]
    );

    return await crypto.subtle.decrypt(
      {
        name: EncryptionConfig.AES.ALGORITHM,
        iv,
        length: EncryptionConfig.AES.KEY_LEN,
        tagLength: EncryptionConfig.AES.TAG_LEN,
      },
      importedKey,
      encryptedData
    );
  } catch (e) {
    console.error(e);
    throw e;
  }
}
