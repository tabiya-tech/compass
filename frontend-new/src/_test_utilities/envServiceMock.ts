jest.mock("src/envService", () => ({
  getFirebaseAPIKey: jest.fn(() => "mock-api-key"),
  getFirebaseDomain: jest.fn(() => "mock-auth-domain"),
  getBackendUrl: jest.fn(() => "mock-backend-url"),
  getApplicationLoginCode: jest.fn(() => ""),
  getApplicationRegistrationCode: jest.fn(() => ""),
  getLoginCodeDisabled: jest.fn(() => "false"),
  getRegistrationDisabled: jest.fn(() => "false"),
  getMetricsEnabled: jest.fn(() => "true"),
  getMetricsConfig: jest.fn(() => ""),
  getCvUploadEnabled: jest.fn(() => "true"),
  getSocialAuthDisabled: jest.fn(() => "false"),
}));
