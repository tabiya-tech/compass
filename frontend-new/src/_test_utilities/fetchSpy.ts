import { ExtendedRequestInit } from "src/utils/customFetch/customFetch";

export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}

// spy on authenticated fetch requests
export function setupAPIServiceSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });

  jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
  return jest.spyOn(require("src/utils/customFetch/customFetch"), "customFetch");
}

export function expectCorrectFetchRequest(fetchSpy: jest.SpyInstance, expectedUrl: string, expectedConfig: ExtendedRequestInit) {
  expect(fetchSpy).toHaveBeenCalledTimes(1);
  const [actualUrl, actualConfig] = fetchSpy.mock.calls[0];
  expect(actualUrl).toBe(expectedUrl);
  expect(actualConfig.method).toBe(expectedConfig.method);

  // Compare headers if they exist
  if(actualConfig.headers) {
    // Handle headers comparison - convert Headers object to plain object if needed
    const actualHeaders = actualConfig.headers instanceof Headers
      ? Object.fromEntries(actualConfig.headers.entries())
      : actualConfig.headers;
    expect(actualHeaders).toEqual(
      expect.objectContaining(expectedConfig.headers),
    );
  }

  // Compare bodies by parsing JSON if they exist
  if (actualConfig.body && expectedConfig.body) {
    let actualBody: unknown;
    let expectedBody: unknown;
    
    try {
      actualBody = JSON.parse(JSON.stringify(actualConfig.body));
      expectedBody = JSON.parse(JSON.stringify(expectedConfig.body));
    } catch (e) {
      console.error('Error parsing request bodies:', e);
      throw new Error('Failed to parse request bodies for comparison');
    }
    expect(actualBody).toEqual(expectedBody);
  }

  // Check any extra properties if provided
  type ExtendedConfigKey = keyof ExtendedRequestInit;
  const extraProperties: ExtendedConfigKey[] = [
    'expectedStatusCode',
    'serviceName',
    'serviceFunction',
    'failureMessage',
    'expectedContentType'
  ] as ExtendedConfigKey[];

  extraProperties.forEach((key) => {
    if (expectedConfig[key] !== undefined) {
      expect(actualConfig[key]).toEqual(expectedConfig[key]);
    }
  });
}

