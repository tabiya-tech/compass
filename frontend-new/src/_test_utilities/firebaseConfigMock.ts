import { AuthProviderIds } from "src/auth/auth.types";

const auth = jest.fn(() => ({
  signInWithCustomToken: jest.fn(),
  signInWithEmailAndPassword: jest.fn(),
  createUserWithEmailAndPassword: jest.fn(),
  signOut: jest.fn(),
  GoogleAuthProvider: { PROVIDER_ID: AuthProviderIds.GOOGLE },
}));

export function mockFirebaseConfig() {
  jest.spyOn(require("src/auth/firebaseConfig"), "auth").mockImplementation(auth);
}
