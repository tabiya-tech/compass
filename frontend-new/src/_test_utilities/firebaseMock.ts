import { AuthProviderIds } from "src/auth/auth.types";

const auth = {
  currentUser: {
    getIdToken: jest.fn(),
    getTokenResult: jest.fn(),
  },
  onAuthStateChanged: jest.fn().mockReturnValue(jest.fn()),
  signInWithPopup: jest.fn(),
  GoogleAuthProvider: { PROVIDER_ID: AuthProviderIds.GOOGLE },
};

const firebase = {
  initializeApp: jest.fn(),
  auth: () => auth,
};

export default firebase;
