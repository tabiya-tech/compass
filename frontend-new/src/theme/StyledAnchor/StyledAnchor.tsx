import { styled } from "@mui/material";
import React from "react";

type Props = Omit<React.HTMLProps<HTMLAnchorElement>, "as" | "ref"> & {
  disabled?: boolean
}

// prevent forwarding the disabled prop to the anchor element
// Because the disabled prop is not a valid prop for an anchor element
// we need to prevent it from being forwarded.
const shouldForwardProp = (prop: string) => prop !== "disabled";

const getValueBasedOnState = (isDisabled: boolean, defaultValue: string, disabledValue: string) => {
  return isDisabled ? disabledValue : defaultValue;
};

const Styled =
  styled("a", { shouldForwardProp })<Props>(({ theme, disabled = false }) => ({
    color: getValueBasedOnState(disabled, theme.palette.tabiyaBlue.main, theme.palette.grey["500"]),
    textDecoration: "underline",
    cursor: getValueBasedOnState(disabled, "pointer", "not-allowed"),
    fontWeight: "bold",
    whiteSpace: "nowrap",
    "&:hover": {
      color: getValueBasedOnState(disabled, theme.palette.tabiyaBlue.light, theme.palette.grey["500"]),
    },
  }));

export const StyledAnchor: React.FC<Props> = ({ onClick, ...props }) => {

  // prevent the onClick event from firing if the anchor is disabled.
  const handleClick = (e: React.MouseEvent<HTMLAnchorElement>) => {
    if (onClick) {
      // if onclick was passed, no reason to redirect.
      e.preventDefault();
    }

    if (!props.disabled) {
      onClick?.(e)
    }
  };

  return <Styled
    aria-disabled={props.disabled}
    onClick={handleClick}
    {...props}
  />;
};
