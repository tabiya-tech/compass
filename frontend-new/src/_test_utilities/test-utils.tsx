// Based on https://testing-library.com/docs/react-testing-library/setup/
import React, { ReactElement, ReactNode } from "react";
import { render, renderHook, RenderHookOptions, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";

// Import the Firebase mock utilities
import "src/_test_utilities/firebaseMock";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/isOnlineProvider/IsOnlineProvider";
jest.mock("firebase/compat/app", () => {
  return {
    initializeApp: jest.fn(),
    auth: jest.fn().mockReturnValue({
      signOut: jest.fn(),
      currentUser: {
        getIdToken: jest.fn(),
        getTokenResult: jest.fn(),
      },
      onAuthStateChanged: jest.fn().mockReturnValue(jest.fn()),
      signInWithPopup: jest.fn(),
      GoogleAuthProvider: { PROVIDER_ID: "google.com" },
    }),
  };
});

jest.mock("firebaseui", () => {
  return {
    auth: {
      AuthUI: {
        start: jest.fn(),
        getInstance: jest.fn().mockReturnValue({
          start: jest.fn(),
          reset: jest.fn(),
        }),
        reset: jest.fn(),
      },
    },
  };
});

jest.mock("@sentry/react", () => ({
  withSentry: (Component: any) => Component,
  getFeedback: jest.fn(),
  withProfiler: (Component: any) => Component,
  wrapCreateBrowserRouter: (Component: any) => Component,
}));

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

function customRenderHook<T>(
  hook: () => T,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { wrapper?: React.ComponentType }
) {
  const CombinedWrapper = ({ children }: { children: ReactNode }) => {
    const Wrapper = options?.wrapper ?? React.Fragment;
    return (
      <AllTheProviders>
        <Wrapper>{children}</Wrapper>
      </AllTheProviders>
    );
  };

  return renderHook(hook, { wrapper: CombinedWrapper, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
export { customRenderHook as renderHook };
