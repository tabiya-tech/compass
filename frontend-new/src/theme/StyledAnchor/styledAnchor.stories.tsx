import type { Meta, StoryObj } from "@storybook/react";

import { StyledAnchor } from "./StyledAnchor";

const meta: Meta<typeof StyledAnchor> = {
  title: "Components/StyledAnchor",
  component: StyledAnchor,
  tags: ["autodocs"],
  argTypes: {},
};

export default meta;
type Story = StoryObj<typeof StyledAnchor>;

export const Shown: Story = {
  args: {
    href: "https://www.tabiya.com",
    children: "Tabiya",
  },
};

export const Disabled: Story = {
  args: {
    disabled: true,
    href: "https://www.tabiya.com",
    children: "Tabiya",
  },
};
