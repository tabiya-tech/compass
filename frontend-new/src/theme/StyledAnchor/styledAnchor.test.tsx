import { render, screen } from "src/_test_utilities/test-utils";
import { StyledAnchor } from "./StyledAnchor";

describe("Styled Anchor tests", () => {
  describe("render tests", () => {
   test("should render correctly default state", () => {
     // GIVEN href and children
     const givenHref = "foo";
     const givenChildren = "bar";

     // AND a data test id.
     const givenDataTestId = "foo-bar";

     // WHEN the component is rendered
     render(<StyledAnchor
       href={givenHref}
       disabled={false}
       data-testid={givenDataTestId}
       children={givenChildren}
     />)

     // THEN it should render an anchor element with the href and children.
     const anchor = screen.getByTestId(givenDataTestId);

     expect(anchor).toBeInTheDocument();
     expect(anchor).toHaveAttribute("href", givenHref);
     expect(anchor).toHaveAttribute("data-testid", givenDataTestId);
     expect(anchor).toHaveTextContent(givenChildren);

     // AND it should not be disabled.
     expect(anchor).not.toHaveAttribute("disabled");
     expect(anchor).toHaveAttribute("aria-disabled", "false");

     // AND it should match the snapshot.
     expect(anchor).toMatchSnapshot();
   })

    test("should render correctly disabled state", () => {
      // GIVEN href and children
      const givenHref = "foo";
      const givenChildren = "bar";

      // AND a data test id.
      const givenDataTestId = "foo-bar";

      // WHEN the component is rendered
      render(<StyledAnchor
        href={givenHref}
        disabled={true}
        data-testid={givenDataTestId}
        children={givenChildren}
      />)

      // THEN it should render an anchor element with the href and children.
      const anchor = screen.getByTestId(givenDataTestId);

      expect(anchor).toBeInTheDocument();
      expect(anchor).toHaveAttribute("href", givenHref);
      expect(anchor).toHaveAttribute("data-testid", givenDataTestId);
      expect(anchor).toHaveTextContent(givenChildren);

      // AND it should be disabled.
      expect(anchor).toHaveAttribute("aria-disabled", "true");

      // AND it should match the snapshot.
      expect(anchor).toMatchSnapshot()
    });
  });

  describe("action tests", () => {
    test("should call onClick when not disabled", () => {
      // GIVEN an onClick handler
      const onClick = jest.fn();

      // AND the component is rendered with the onClick handler and not disabled.
      render(<StyledAnchor
        href="foo"
        disabled={false}
        onClick={onClick}
      />);

      // WHEN the anchor is clicked
      screen.getByRole("link").click();

      // THEN the onClick handler should be called
      expect(onClick).toHaveBeenCalled();
    })

    test("should not call onClick when disabled", () => {
      // GIVEN an onClick handler
      const onClick = jest.fn();

      // AND the component is rendered with the onClick handler and disabled.
      render(<StyledAnchor
        href="foo"
        disabled={true}
        onClick={onClick}
      />);

      // WHEN the anchor is clicked
      screen.getByRole("link").click();

      // THEN the onClick handler should not be called
      expect(onClick).not.toHaveBeenCalled();
    })
  })
});
