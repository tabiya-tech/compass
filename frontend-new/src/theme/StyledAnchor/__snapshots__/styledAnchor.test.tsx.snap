// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Styled Anchor tests render tests should render correctly default state 1`] = `
<a
  aria-disabled="false"
  class="css-1dwh1c3"
  data-testid="foo-bar"
  href="foo"
>
  bar
</a>
`;

exports[`Styled Anchor tests render tests should render correctly disabled state 1`] = `
<a
  aria-disabled="true"
  class="css-1w3ikts"
  data-testid="foo-bar"
  href="foo"
>
  bar
</a>
`;
